packages/contracts/contracts/factory/PackageInstaller.sol

```diff
@@ -0,0 +1,92 @@
+/*
+ * SPDX-License-Identifier:    MIT
+ */
+
+pragma solidity 0.8.10;
+
+import "../APM/IPackage.sol";
+import "./../core/DAO.sol";
+import "../utils/UncheckedIncrement.sol";
+
+/// @title PackageInstaller to install packages on a DAO.
+/// @author Sarkawt Noori - Aragon Association - 2022
+/// @notice This contract is used to create/deploy new packages and instaling them on a DAO.
+contract PackageInstaller {
+    address public tokenFactory;
+
+    /**
+    NOTE: better aproach might be to have a permission ops like:
+
+    struct PermissionOp {
+        PermissionType permissionType;
+        bytes32[] roles;
+        address where;
+        address who;
+    }
+
+    to be used inside Packages like:
+
+    struct Package {
+        address factoryAddress; // package deployer (factory) address, hopefully from APM
+        PermissionOp[] PermissionOps; 
+        bytes args; // pre-determined value for stting up the package
+    }
+
+    for futher improvment and relationship between apps, utilizing AppArray and indexes
+
+     */
+    struct Package {
+        bool isNewTokenNeeded;
+        address factoryAddress; // package deployer (factory) address, hopefully from APM
+        bytes32[] PackagePermissions; // to be granted to DAO
+        bytes32[] DAOPermissions; // Dao permission to be granted to package like: exec_role
+        bytes args; // pre-determined value for stting up the package
+    }
+
+    event PackageInstalled(address indexed dao, address indexed packageAddress);
+
+    error NoRootRole();
+
+    constructor(address _tokenFactory) {
+        tokenFactory = _tokenFactory;
+    }
+
+    function installPckagesOnExistingDAO(DAO dao, Package[] calldata packages) external {
+        if (dao.hasPermission(address(this), address(dao), dao.ROOT_ROLE(), bytes("0x00"))) revert NoRootRole();
+        for (uint256 i; i < packages.length; i = _uncheckedIncrement(i)) {
+            installPackage(dao, packages[i]);
+        }
+        dao.revoke(address(dao), address(this), dao.ROOT_ROLE());
+    }
+
+    function installPackage(DAO dao, Package calldata packages) internal returns (address app) {
+        // revoke root from TokenFactory
+        if (packages.isNewTokenNeeded) {
+            dao.grant(address(dao), address(tokenFactory), dao.ROOT_ROLE());
+        }
+
+        // deploy new packaes for Dao
+        app = IPackage(packages.factoryAddress).deploy(address(dao), packages.args);
+
+        // revoke root from TokenFactory
+        if (packages.isNewTokenNeeded) {
+            dao.revoke(address(dao), address(tokenFactory), dao.ROOT_ROLE());
+        }
+
```

Comment:
Could this be skipped/moved?

Reply:
may be the isNewTokenNeeded is not a good naming , it changed it to isGovernanceTokenNeeded....
so there should be a way for the Plugin to tell the PlugIn installer if it needs to set permission for token factory or not...

do u you think there is a better way ?